src/normalize.js

```diff
@@ -1,87 +1,385 @@
+    parent: parentNode.id,
+    children: [],
+    [attributeName]: text,
+    internal: {
+      type: _.camelCase(`${parentNode.internal.type}-${attributeName}-TextNode`),
+      mediaType: `text/markdown`,
+      content: text,
+      contentDigest: createContentDigest(text),
+    },
+  };
 
-    return;
-  }
+  return textNode;
+};
 
-  if (isObject(item)) {
-    for (const key in item) {
-      await extractFields(item[key], ctx);
-    }
+/**
+ * Returns an array of the main node and children nodes to create
+ * @param {Object} entity the main entry
+ * @param {String} nodeType the name of the main node
+ * @param {Object} ctx object of gatsby functions
+ * @param {String} uid the main schema uid
+ * @returns {Object[]} array of nodes to create
+ */
+export const createNodes = (entity, nodeType, ctx, uid) => {
+  const nodes = [];
+
+  const { schemas, createNodeId, createContentDigest, getNode } = ctx;
+
+  let entryNode = {
+    id: createNodeId(`${nodeType}-${entity.id}`),
+    strapi_id: entity.id,
+    parent: null,
+    children: [],
+    internal: {
+      type: nodeType,
+      content: JSON.stringify(entity),
+      contentDigest: createContentDigest(entity),
+    },
+  };
+
+  const schema = getContentTypeSchema(schemas, uid);
+
+  for (const attributeName of Object.keys(entity)) {
+    const value = entity[attributeName];
+
+    const attribute = schema.schema.attributes[attributeName];
+    const type = _.get(attribute, 'type', null);
+
+    if (value) {
+      // Add support for dynamic zones
+      if (type === 'dynamiczone') {
+        value.forEach((v) => {
+          const componentSchema = getContentTypeSchema(schemas, v.strapiComponent);
+          const componentNode = `StrapiComponent${_.upperFirst(_.camelCase(componentSchema.uid))}`;
+
+          const valueNodes = _.flatten(createNodes(v, componentNode, ctx, v.strapiComponent));
+          const compoNodeIds = valueNodes
+            .filter(({ internal }) => internal.type === componentNode)
+            .map(({ id }) => id);
+
+          entity[`${attributeName}___NODE`] = [
+            ...(entity[`${attributeName}___NODE`] || []),
+            ...compoNodeIds,
+          ];
+
+          valueNodes.forEach((n) => {
+            nodes.push(n);
+          });
+        });
+
+        delete entity[attributeName];
+      }
+
+      if (type === 'relation') {
+        // Create type for the first level of relations, otherwise the user should fetch the other content type
+        // to link them
+        const config = {
+          schemas,
+          createContentDigest,
+          createNodeId,
+          parentNode: entryNode,
+          parentNodeType: nodeType,
+          attributeName,
+          targetSchemaUid: attribute.target,
+        };
+
+        if (Array.isArray(value)) {
+          const relationNodes = value.map((relation) => prepareRelationNode(relation, config));
+          entity[`${attributeName}___NODE`] = relationNodes.map(({ id }) => id);
```

Comment:
please do not use node mutations. Data could be lost. In gatsby v4 they are using LMDB store. https://www.gatsbyjs.com/docs/how-to/local-development/debugging-missing-data/#mutating-a-node-in-oncreatenode. 

> Direct node mutations in various API lifecycles are not persisted anymore. In previous Gatsby versions it did work because source of truth for the data layer was directly in Node.js memory, so mutating node was in fact mutating source of truth. Now Gatsby edits data it receives from the database, so unless it explicitly upserts data to this database after edits, those edits will not be persisted (even for same the same build).

Reply:
e.g. I mentioned that case in [here](https://github.com/strapi/gatsby-source-strapi/pull/122#issuecomment-996759066).

BTW. Do you plan to support in-text images as I did in my code?

Reply:
I see I think I would have to change the entire implementation in order to do see, won't I?

`createNodeField` is called on a created node so I would have to first create the parent node then the children ones it seems doable.

I have looked at other source-plugins and most of them seems to be using the `___NODE` api instead of the createNodeField one so I guess it is still allowed to do so?

WDYT?

Reply:
I'm aware that many source plugins still doing like this but look like avoiding mutations will be more future proof. 
Even there is a tutorial [how to migrate a plugin to v4](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/#2-data-mutations-need-to-happen-during-sourcenodes-or-oncreatenode). 

My concern is that Gatsby will block the ability to mutate nodes in the future. Or it could cause some side effects if they go with more parallelization. 

And about examples, they point to see how it should be done by looking at `gatsby-source-wordpress` or `gatsby-source-shopify`. Maybe it could be helpful.

Reply:
> e.g. I mentioned that case in [here](https://github.com/strapi/gatsby-source-strapi/pull/122#issuecomment-996759066).
> 
> BTW. Do you plan to support in-text images as I did in my code?

Yes as soon as I have handled delta fetches and that I support gatsby I will extract images for markdown,  @timbrandin also did a good implementation for this [here](https://github.com/relate-app/gatsby-source-strapi/blob/9a425597b25707c41750e3454e0ec1c77cb2e283/helpers.js#L124-L144)

Reply:
good to hear ❤️